integrations/slack/src/actions/action-wrapper.ts

```diff
@@ -0,0 +1,72 @@
+import { RuntimeError } from '@botpress/sdk'
+import { WebClient, ErrorCode } from '@slack/web-api'
+import { getAccessToken } from '../misc/utils'
+import { IntegrationProps } from '.botpress'
+
+type Actions = {
+  [K in keyof IntegrationProps['actions']]: IntegrationProps['actions'][K]
+}
+type ActionInjections = { slackClient: WebClient }
+type ActionWithInjections = {
+  [K in keyof Actions]: (
+    props: Parameters<Actions[K]>[0] & ActionInjections,
+    input: Parameters<Actions[K]>[0]['input']
+  ) => ReturnType<Actions[K]>
+}
+
+const tryCatch = async <T>(fn: () => Promise<T>, errorMessage: string): Promise<T> => {
+  try {
+    return await fn()
+  } catch (thrown: unknown) {
+    const error = thrown instanceof Error ? thrown : new Error(`${thrown}`)
+    console.error(`Error: ${errorMessage}`, error)
```

Comment:
If this error is mean to be surfaced in the bot logs we should use `logger.forBot().error()` instead. If not and we just want the error to be shown in the integration logs, then just rethrowing the error is enough.

Also, the `Error: ` prefix in the error message might be unnecessary? As the error logs are already visually distinguished in the logs UI.

Reply:
The intent is to throw a RuntimeError so that the bot owner receives the error, but I also want the original error to get logged in Grafana, hence the `console.error`. If I were to just re-throw the original error, it wouldn't appear in either the integration logs or the bot logs because it wouldn't be a RuntimeError.

Thus, we first log the original error, then we redact sensitive information from it and throw the result as a RuntimeError.

Reply:
AFAIK we don't send integration logs to Grafana, as they're natively stored in CloudWatch and we have a lambda that subscribes to the CW logs and stores the logs in DynamoDB, where the logs UI retrieves the logs from for the bots and integrations.

You could check with @spg who built the integration logging architecture to confirm the best/correct way to implement what you're trying to achieve, but if I'm not mistaken and following the existing pattern in the other integrations, you just need to surface the redacted error message to the bot using `logger.forBot().error()` and then simply rethrow the original error so it's surfaced in the integration logs.

Reply:
Thanks for the explanations. I've done some testing and here's what I found:

- `console.log`, `console.error`, etc:
  - not logged in grafana
  - not logged in bot logs
  - not logged in the studio
  - logged in integration logs
- `logger.forBot()...`:
  - not logged in grafana
  - logged in bot logs
  - not logged in the studio
  - logged in integration logs
- throwing any error aside from `RuntimeError`:
  - not logged in grafana
  - logged as a generic error in the bot logs: `An unexpected error occurred in the integration. Bot owners: Check logs for more informations. Integration owners: throw a RuntimeError to return a custom error message instead.`
  - logged as a generic error in the studio
  - logged in the integration logs
- throwing a `RuntimeError`:
  - logged in grafana
  - logged in bot logs
  - logged in the studio
  - logged in integration logs

---

What I'm looking to accomplish with the error handler is the following:
- the redacted error should appear in the studio when bot owners are trying out their bots
- the original error should be available to the integration owner (Botpress), so either logged in grafana or in the integration logs. It should not be visible to the bot owner.

Thus, it appears the only way to achieve this is to `console.error` the original error and to throw a `RuntimeError` with the redacted error message.

Reply:
OK, thanks for clarifying what's being intended here. Might be worth adding a comment right above this `console.error` briefly explaining why we use this behavior, to avoid someone inadvertently removing it in the future as I believe most of the other integrations don't use this pattern (yet).